src/state.ts

```diff
@@ -23,62 +23,120 @@ export type TransitionEvent<TContext = unknown, TEvent = unknown> = (
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
 }
 
-export interface InitialAutoTransitionState<
+interface InitialWithAutoTransitionAndFinalFalseState<
 	TContext = unknown,
 	TEvent = unknown,
 > {
 	id: StateIdentifier;
-	initial: boolean;
-	autoTransition: boolean;
+	initial: true;
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
+	final: false;
 }
 
-export interface FinalState<TContext = unknown, TEvent = unknown> {
+export interface TransitoryState<TContext = unknown, TEvent = unknown> {
 	id: StateIdentifier;
-	final: boolean;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
-	onFinal?: OnFinalHook<TContext>;
-}
+	transitionTo: TransitionToHook<TContext, TEvent>;
 
-export type GenericState<TContext = unknown, TEvent = unknown> =
-	| GuardState
-	| AutoTransitionState;
+	initial?: boolean;
+	autoTransition?: false;
+	transitionGuard?: TransitionEvent<TContext, TEvent>;
+}
 
-export interface GuardState<TContext = unknown, TEvent = unknown> {
+interface TransitoryWithFinalFalseState<TContext = unknown, TEvent = unknown> {
 	id: StateIdentifier;
+	transitionTo: TransitionToHook<TContext, TEvent>;
+	final: false;
+
+	initial?: boolean;
+	autoTransition?: false;
 	transitionGuard?: TransitionEvent<TContext, TEvent>;
+}
+
+interface TransitoryWithAutoTransitionState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
+	id: StateIdentifier;
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
+
+	initial?: boolean;
 }
 
-export interface AutoTransitionState<TContext = unknown, TEvent = unknown> {
+interface TransitoryWithAutoTransitionAndFinalFalseState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
 	id: StateIdentifier;
-	autoTransition: boolean;
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
+	final: false;
+
+	initial?: boolean;
+}
+
+export interface FinalState<TContext = unknown, TEvent = unknown> {
+	id: StateIdentifier;
+	final: true;
+
+	initial?: false;
+	onFinal?: OnFinalHook<TContext>;
+}
+
+export interface GenericCallbackState<TContext = unknown, TEvent = unknown> {
 	onEntry?: OnEntryHook<TContext, TEvent>;
 	onExit?: OnExitHook<TContext, TEvent>;
 }
 
-export type State<TContext = unknown, TEvent = unknown> =
-	| InitialGuardState
-	| InitialAutoTransitionState
-	| FinalState
-	| GuardState
-	| AutoTransitionState;
+export type AutoTransitionState<TContext = unknown, TEvent = unknown> =
+	| InitialWithAutoTransitionState
+	| InitialWithAutoTransitionAndFinalFalseState
+	| TransitoryWithAutoTransitionState
+	| TransitoryWithAutoTransitionAndFinalFalseState;
+
+export type GuardState<TContext = unknown, TEvent = unknown> = State<
```

Comment:
how is `GuardState` respecting the constraint that `autoTransition` is false?

Reply:
You are right, i will fix that